hawc/tools/tables/sot.py

```diff
@@ -0,0 +1,174 @@
+
+    def get_cell(self, selection: pd.DataFrame) -> BaseCell:
+        return getattr(self, f"get_{self.value}")(selection)
+
+
+class Subheader(BaseModel):
+    label: str
+    start: conint(ge=0)
+    length: conint(ge=1)
+
+
+class Column(BaseModel):
+    label: str
+    attribute: AttributeChoices
+    metric: Optional[int]
+    key: str
+
+
+class Row(BaseModel):
+    id: int
+    type: str
+    customized: List
+
+
+class JudgmentCell(BaseCell):
+    judgment: int
+
+    def to_docx(self, parser: QuillParser, block):
+        text = SCORE_SYMBOLS[self.judgment]
+        color = SCORE_SHADES[self.judgment].strip("#")
+        color_background(block, color)
+        return parser.feed(tag_wrapper(text, "p"), block)
+
+
+class StudyOutcomeTable(BaseTable):
+    assessment_id: int
+    subheaders: List[Subheader]
+    cell_columns: List[Column] = Field([], alias="columns")
+    cell_rows: List[Row] = Field([], alias="rows")
+
+    def _subheaders_group(self):
+        cells = []
+        for subheader in self.subheaders:
+            html = tag_wrapper(subheader.label, "p", "strong")
+            cells.append(
+                GenericCell.parse_args(True, 0, subheader.start, 1, subheader.length, html)
+            )
+        return BaseCellGroup.construct(cells=cells)
+
+    def _columns_group(self):
+        cells = []
+        for i, col in enumerate(self.cell_columns):
+            html = tag_wrapper(col.label, "p", "strong")
+            cells.append(GenericCell.parse_args(True, 0, i, 1, 1, html))
+        return BaseCellGroup.construct(cells=cells)
+
+    def _set_override(self, final_scores: pd.DataFrame, cell: BaseCell, row: Row, column: Column):
+        for custom in [custom for custom in row.customized if custom["key"] == column.key]:
+            if "html" in custom:
+                cell.quill_text = custom["html"]
+            elif "score_id" in custom:
+                values = final_scores.loc[final_scores["score_id"] == custom["score_id"]][
+                    "score_score"
+                ].values
+                cell.judgment = 0 if values.size == 0 else values[0]
+
+    def _get_selection(
+        self, bioassay_data: pd.DataFrame, final_scores: pd.DataFrame, row: Row, column: Column
+    ) -> pd.DataFrame:
+        if column.attribute.value == "rob_score":
+            return final_scores.loc[
+                (final_scores["study_id"] == row.id)
+                & (final_scores["metric_id"] == column.metric)
+                & (final_scores["content_type_id"].isnull())
+            ]
+        else:
+            return bioassay_data.loc[bioassay_data["study id"] == row.id]
+
+    def _filter_rows(self, bioassay_data: pd.DataFrame) -> List[Row]:
+        return [row for row in self.cell_rows if row.id in bioassay_data["study id"].values]
+
+    def _rows_group(self):
+        cells = []
+
+        study_ids = {row.id for row in self.cell_rows if row.type == "study"}
+        final_scores = pd.DataFrame.from_records(
```

Comment:
you might be able to set the index for the study_id and metric_id here, that may make the lookups faster, if those keys are always reuqired which i think they are

Reply:
Overrides are also included in the dataframe, so `study_id` and `metric_id` won't make unique indices.

Reply:
unlike databases, i dont htink indexes in data frames have to be unique. https://stackoverflow.com/questions/16626058/what-is-the-performance-impact-of-non-unique-indexes-in-pandas

Reply:
Good to know! I'll give it a shot. I was deterred by the ominous wording under `verify_integrity`: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html

Reply:
I added the index in my latest commit fd34bba7c47451f256b317d826feb6ffe53c18ce. There was a bit of a tradeoff, which was kind of alluded to in your SO link; if the index is unique, then `loc` will return a series, and if its not unique, then it will return a dataframe. I also had to take into account `KeyError`s when indexing into the dataframe.

Reply:
Do you think this tradeoff is worth it? If so, we can keep, otherwise I'm fine with reverting.